File: CSharp/Clipper2Lib.Tests/Tests1/Tests/TestPolygons.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Clipper2Lib.UnitTests
{

  [TestClass]
  public class TestPolygons
  {

    private static bool IsInList(int num, int[] list)
    {
      foreach (int i in list) if (i == num) return true;
      return false;
    }

    [TestMethod]
    public void TestClosedPaths()
    {
      int testNum = 0;
      while (true)
      {
        testNum++;
        Clipper64 c64 = new();
        Paths64 subj = new(), subj_open = new (), clip = new();
        Paths64 solution = new(), solution_open = new();

        if (!ClipperFileIO.LoadTestNum("..\\..\\..\\..\\..\\..\\Tests\\Polygons.txt",
          testNum, subj, subj_open, clip, out ClipType clipType, out FillRule fillrule, 
          out long storedArea, out int storedCount, out _))
        {          
          Assert.IsTrue(testNum > 180, string.Format("Loading test polygon {0} failed.", testNum));
          break;
        }

        c64.AddSubject(subj);
        c64.AddOpenSubject(subj_open);
        c64.AddClip(clip);
        c64.Execute(clipType, fillrule, solution, solution_open);
        int measuredCount = solution.Count;
        long measuredArea = (long)Clipper.Area(solution);
        int countDiff = storedCount > 0 ? Math.Abs(storedCount - measuredCount) : 0;
        long areaDiff = storedArea > 0 ? Math.Abs(storedArea - measuredArea) : 0;
        double areaDiffRatio = storedArea <= 0 ? 0 : (double) areaDiff / storedArea;

        // check polygon counts
        if (storedCount > 0)
        {
          if (IsInList(testNum, new int[] { 140, 150, 165, 166, 172, 173, 176, 177, 179 }))
          {
            Assert.IsTrue(countDiff <= 9);
          }
          else if (testNum >= 120)
          {
            Assert.IsTrue(countDiff <= 6);
          }
          else if (IsInList(testNum, new int[] { 27, 121, 126 }))
            Assert.IsTrue(countDiff <= 2);
          else if (IsInList(testNum, new int[] { 23, 37, 43, 45, 87, 102, 111, 118, 119 }))
            Assert.IsTrue(countDiff <= 1);
          else
            Assert.IsTrue(countDiff == 0);
        }

        // check polygon areas
        if (storedArea > 0)
        {
          if (IsInList(testNum, new int[] { 19, 22, 23, 24 }))
            Assert.IsTrue(areaDiffRatio <= 0.5);
          else if (testNum == 193)
            Assert.IsTrue(areaDiffRatio <= 0.25);
          else if (testNum == 63)
            Assert.IsTrue(areaDiffRatio <= 0.1);
          else if (testNum == 16)
            Assert.IsTrue(areaDiffRatio <= 0.075);
          else if (IsInList(testNum, new int[] { 15, 26 }))
            Assert.IsTrue(areaDiffRatio <= 0.05);
          else if (IsInList(testNum, new int[] { 52, 53, 54, 59, 60, 64, 117, 118, 119, 184 }))
            Assert.IsTrue(areaDiffRatio <= 0.02);
          else
            Assert.IsTrue(areaDiffRatio <= 0.01);
        }

      } //bottom of num loop

    }
  }
}

File: CPP/Tests/TestPolygons.cpp
#include <gtest/gtest.h>
#include "clipper2/clipper.h"
#include "ClipFileLoad.h"

using Clipper2Lib::Scalar;

inline Clipper2Lib::PathS MakeRandomPath(int width, int height, unsigned vertCnt)
{
  Clipper2Lib::PathS result;
  result.reserve(vertCnt);
  for (unsigned i = 0; i < vertCnt; ++i)
    result.push_back(Clipper2Lib::PointS(Scalar(rand()) / RAND_MAX * width, Scalar(rand()) / RAND_MAX * height));
  return result;
}
template <size_t N>
inline bool IsInList(int num, const int (&intArray)[N])
{
  const int* list = &intArray[0];
  for (int cnt = N; cnt; --cnt)
    if (num == *list++) return true;
  return false;
}
TEST(Clipper2Tests, TestMultiplePolygons)
{
  std::ifstream ifs("Polygons.txt");
  ASSERT_TRUE(ifs.good());
  const int start_num = 1;
  const int end_num = 1000;
  int test_number = start_num;
  while (test_number <= end_num)
  {
    Clipper2Lib::PathsI subject, subject_open, clip;
    Clipper2Lib::PathsI solution, solution_open;
    Clipper2Lib::ClipType ct;
    Clipper2Lib::FillRule fr;
    Integer stored_area, stored_count;
    if (!LoadTestNum(ifs, test_number,
      subject, subject_open, clip, stored_area, stored_count, ct, fr)) break;
    // check PathsI solutions
    Clipper2Lib::ClipperI c;
    c.AddSubject(subject);
    c.AddOpenSubject(subject_open);
    c.AddClip(clip);
    c.Execute(ct, fr, solution, solution_open);
    const Integer measured_area = static_cast<Integer>(Area(solution));
    const Integer measured_count = static_cast<Integer>(solution.size() + solution_open.size());
    // check the polytree variant too
    Clipper2Lib::PolyTreeI solution_polytree;
    Clipper2Lib::PathsI solution_polytree_open;
    Clipper2Lib::ClipperI clipper_polytree;
    clipper_polytree.AddSubject(subject);
    clipper_polytree.AddOpenSubject(subject_open);
    clipper_polytree.AddClip(clip);
    clipper_polytree.Execute(ct, fr, solution_polytree, solution_polytree_open);
    const Integer measured_area_polytree =
      static_cast<Integer>(solution_polytree.Area());
    const auto solution_polytree_paths = PolyTreeToPathsI(solution_polytree);
    const Integer measured_count_polytree =
      static_cast<Integer>(solution_polytree_paths.size());
    // check polygon counts
    if (stored_count <= 0)
      ; // skip count
    else if (IsInList(test_number, { 120, 121, 130, 138,
      140, 148, 163, 165, 166, 167, 168, 172, 175, 178, 180 }))
      EXPECT_NEAR(measured_count, stored_count, 5) << " in test " << test_number;
    else if (IsInList(test_number, { 27, 181 }))
      EXPECT_NEAR(measured_count, stored_count, 2) << " in test " << test_number;
    else if (test_number >= 120 && test_number <= 184)
      EXPECT_NEAR(measured_count, stored_count, 2) << " in test " << test_number;
    else if (IsInList(test_number, { 23, 45, 87, 102, 111, 113, 191 }))
      EXPECT_NEAR(measured_count, stored_count, 1) << " in test " << test_number;
    else
      EXPECT_EQ(measured_count, stored_count) << " in test " << test_number;
    // check polygon areas
    if (stored_area <= 0)
      ; // skip area
    else if (IsInList(test_number, { 19, 22, 23, 24 }))
      EXPECT_NEAR(measured_area, stored_area, 0.5 * measured_area) << " in test " << test_number;
    else if (test_number == 193)
      EXPECT_NEAR(measured_area, stored_area, 0.2 * measured_area) << " in test " << test_number;
    else if (test_number == 63)
      EXPECT_NEAR(measured_area, stored_area, 0.1 * measured_area) << " in test " << test_number;
    else if (test_number == 16)
      EXPECT_NEAR(measured_area, stored_area, 0.075 * measured_area) << " in test " << test_number;
    else if (test_number == 26)
      EXPECT_NEAR(measured_area, stored_area, 0.05 * measured_area) << " in test " << test_number;
    else if (IsInList(test_number, { 15, 52, 53, 54, 59, 60, 64, 117, 119, 184 }))
      EXPECT_NEAR(measured_area, stored_area, 0.02 * measured_area) << " in test " << test_number;
    else
      EXPECT_NEAR(measured_area, stored_area, 0.01 * measured_area) << " in test " << test_number;
    EXPECT_EQ(measured_count, measured_count_polytree)
      << " in test " << test_number;
    EXPECT_EQ(measured_area, measured_area_polytree)
      << " in test " << test_number;
    ++test_number;
  }
  //EXPECT_GE(test_number, 188);
  Clipper2Lib::PathsS subjd, clipd, solutiond;
  Clipper2Lib::FillRule frd = Clipper2Lib::FillRule::NonZero;
}
TEST(Clipper2Tests, TestHorzSpikes) //#720
{
  Clipper2Lib::PathsI paths = {
    Clipper2Lib::MakePath({1600,0, 1600,100, 2050,100, 2050,300, 450,300, 450, 0}),
    Clipper2Lib::MakePath({1800,200, 1800,100, 1600,100, 2000,100, 2000,200}) };
  std::cout << paths << std::endl;
  Clipper2Lib::ClipperI c;
  c.AddSubject(paths);
  c.Execute(Clipper2Lib::ClipType::Union, Clipper2Lib::FillRule::NonZero, paths);
  EXPECT_GE(paths.size(), 1);
}

TEST(Clipper2Tests, TestCollinearOnMacOs) //#777
{
  Clipper2Lib::PathsI subject;
  subject.push_back(Clipper2Lib::MakePath({ 0, -453054451,0, -433253797,-455550000, 0 }));
  subject.push_back(Clipper2Lib::MakePath({ 0, -433253797,0, 0,-455550000, 0 }));
  Clipper2Lib::ClipperI clipper;
  clipper.PreserveCollinear(false);
  clipper.AddSubject(subject);
  Clipper2Lib::PathsI solution;
  clipper.Execute(Clipper2Lib::ClipType::Union, Clipper2Lib::FillRule::NonZero, solution);
  ASSERT_EQ(solution.size(), 1);
  EXPECT_EQ(solution[0].size(), 3);
  EXPECT_EQ(IsPositive(subject[0]), IsPositive(solution[0]));
}